dll/win32/msgina/shutdown.c

```diff
@@ -538,6 +561,89 @@ ShutdownDialog(
     return ret;
 }
 
+/*************************************************************************
+ * SetRegistryToBootInNTNativeMode
+ *
+ * NOTES
+ *     Used to append the NT Native shell executable to the BootExecute
+ *     registry value then triggers a restart in order to boot into it.
+ */
+VOID SetRegistryToBootInNTNativeMode(VOID)
+{
+    // prepare the registry for booting into the NT Native Shell
+    LONG lRet;
+    HKEY hKey;
+    DWORD dwType = 0, dwCurrentValueSize = 0;
+    WCHAR *pszCurrentValue = NULL, *pszFinalValue = NULL;
+    UINT cchLatest = 0;
+    const UINT wcharSize = sizeof(WCHAR);
+#define SHELL_ARG_LENGTH 22
+    // '0' will be replaced with null terminators
+    WCHAR pszNativeShellArg[SHELL_ARG_LENGTH] = L"0native Hello World!0";
+
+    lRet = RegCreateKeyExW(
+        HKEY_LOCAL_MACHINE, L"SYSTEM\\CurrentControlSet\\Control\\Session Manager", 0, NULL, REG_OPTION_NON_VOLATILE,
+        KEY_ALL_ACCESS, NULL, &hKey, NULL);
+
+    if (lRet != ERROR_SUCCESS)
+    {
+        TRACE("Unable to open the Session Manager key, error %d", GetLastError());
+        return;
+    }
+
+    lRet = RegQueryValueEx(hKey, L"BootExecute", NULL, &dwType, NULL, &dwCurrentValueSize);
+
+    if (lRet != ERROR_SUCCESS || dwType != REG_MULTI_SZ)
+    {
+        TRACE("Unable to grab BootExecute, error %d", GetLastError());
+        return;
+    }
+
+    const UINT uInitialStringLength = lstrlen(pszCurrentValue);
+    pszCurrentValue = (WCHAR *)HeapAlloc(GetProcessHeap(), 0, dwCurrentValueSize);
+    if (!pszCurrentValue)
+    {
+        TRACE("HeapAlloc failed to allocate %d bytes", dwCurrentValueSize);
+        return;
+    }
+    memset(pszCurrentValue, '\0', uInitialStringLength);
+
+    lRet = RegQueryValueExW(hKey, L"BootExecute", NULL, &dwType, (LPBYTE)pszCurrentValue, &dwCurrentValueSize);
+
+    cchLatest = dwCurrentValueSize - wcharSize + SHELL_ARG_LENGTH * wcharSize;
+
+    // check if we haven't set the key already
+    if (NULL == wcsstr(pszCurrentValue, pszNativeShellArg))
+    {
+        RegCloseKey(hKey);
+        HeapFree(GetProcessHeap(), 0, pszCurrentValue);
+        return;
+    }
+
+    pszFinalValue = (WCHAR *)HeapAlloc(GetProcessHeap(), 0, cchLatest);
+    if (!pszFinalValue)
+    {
+        TRACE("HeapAlloc failed to allocate %d bytes\n", cchLatest);
+        goto Cleanup;
+    }
+    memset(pszFinalValue, '\0', (cchLatest / wcharSize));
+
+    wcscpy(&pszFinalValue[0], &pszCurrentValue[0]);
+    lstrcat((LPWSTR)pszFinalValue, pszNativeShellArg);
+
+    // MULTI_SZ uses null terminators as string delimiters
+    *(pszFinalValue + uInitialStringLength) = '\0';
+    *(pszFinalValue + uInitialStringLength + SHELL_ARG_LENGTH - 2) = '\0';
```

Comment:
Please let the compiler handle stuff like this (and the magic 22 ;)
```
    WCHAR pszNativeShellArg1[] = L"0native Hello World!0";
    WCHAR pszNativeShellArg2[] = L"\0native Hello World!\0";

    printf("1: %u\n",_countof(pszNativeShellArg1));
    printf("2: %u\n",_countof(pszNativeShellArg2));
```
Will output:
```
1: 22
2: 22
```